File: 01-one-variable-equations/04-newton-rhapson/newton_raphson.m

%{
Newton-Raphson method

@param x0 - initial value
@param func - lambda function representing the function to work
@param errorFunc - lambda function to compute the error
@param tol - tolerance of the algorithm / error
@param maxIter - max number of iterations

@return x - approximation of x
@return iterations - number of iterations
@return currentError - error
%}

clear;
close all;

% Only when running on PC
%pkg load symbolic;


function [xAprox, iterations, currentError] = newtonRaphson (x0, func, errorFunc, tol, maxIter)
    % Initial value of x
    xAprox = x0;
    % Current iterations
    iterations = 0;

    % Convert function to symbolic programming
    syms f(x);
    f(x) = func;
    % Derivate function
    df = diff(f);
    % Current error
    currentError = x0;
    
    % Repeat until x approaches the root of the function
    for _=1:maxIter
        % Save previous value
        xk_1 = xAprox;
        # Evaluate function
        fxk_1 = func(xk_1);
        # Evaluate in df/fx
        dfxk_1 = double(df(xk_1));

        % Update x value and the current iteration
        xAprox = xk_1 - fxk_1 / dfxk_1;

        % New error
        currentError = errorFunc(xAprox, xk_1);

        # Show info
        if iterations > 0
            printf("%d\t%f\t%f\t%f\t%f\t-\n",
                    iterations + 1, xk_1, fxk_1, dfxk_1, xAprox);
        else
            printf("%d\t%f\t%f\t%f\t%f\t%f\n",
                    iterations + 1, xk_1, fxk_1, dfxk_1, xAprox, currentError);
        endif

        iterations += 1;

        % Check error
        if (currentError < tol)
            break;
        endif
    endfor

    disp("Derivada: ");
    disp(df);

    return;
endfunction


% Main
% Initial value
x0 = 1.5;
% Tolerance
tol = 0.001;
% Max number of iterations
maxIter = 1000;
% Function to work
func = @(x) x^2 - cos(x) - 1;
errorFunc = @(xi, xi_1) abs(xi - xi_1) / abs (xi);
% Call method
[xi, iterations, finalError] = newtonRaphson (x0, func, errorFunc, tol, maxIter);

% Display result
printf("La aproximación de x es %f después de %d iteraciones con un error de %f\n", xi, iterations, finalError);

File: 01-one-variable-equations/04-newton-rhapson/newton_raphson.py
from math import exp, sin, cos, tan, pi
from scipy import misc
from typing import Tuple


def newton_raphson_method(x0: float,
                          func=None,
                          error_func=None,
                          error: float = 1e-10,
                          _iter: int = 9e10) -> Tuple[float, float, int]:
    """Newton-Raphson method

    Params:
    ------------------------------------------------------------------
        - x0: float.
            initial value of x
        - func
            lambda function representing the function to work
        - error_func
            lambda function to compute the error
        - error: float
            error value
        - _iter: int
            number of iterations

    Returns
    ------------------------------------------------------------------
        - Approximation of x
        - Error
        - Number of iterations
    """
    # Valor incial de x
    xk: float = x0
    i: int = 0

    # Current error of the approximation
    current_error: float = 1.0

    while (error < current_error and i < _iter):
        # Get x in the iteration k - 1
        xk_1 = xk
        # Evaluating the function
        fxk_1 = func(xk)

        # Derivating and evaluating the function
        dfxk_1 = misc.derivative(func, xk_1, dx=1e-6)

        # Current approximation of x
        xk = xk - fxk_1 / dfxk_1

        # New error
        current_error = error_func(xk, xk_1)

        # Show info
        print(i + 1, xk_1, fxk_1, dfxk_1, xk,
              current_error if i > 0 else "-", sep="\t")

        i += 1

    return xk, current_error, i


if __name__ == '__main__':
    # Initial value
    x0 = 1.5
    # Tolerance / error
    tol = 0.0001
    # Function to work
    func = lambda x: pow(x, 2) - cos(x) - 1
    # Error function
    error_func = lambda xi, xi_1: abs(xi - xi_1) / abs(xi)

    # Compute with the false position method
    xi, error, i = newton_raphson_method(x0, func, error_func, tol, _iter=5)
    print(
        f"La aproximacion de x es {xi} y tomo {i} iteraciones con un error de {error}")
